src/data/matrix/notation.lean

```diff
@@ -130,6 +130,130 @@ cons_val_succ x u 0
   vec_cons x u i = x :=
 by { fin_cases i, refl }
 
+/-! ### Numeral (`bit0` and `bit1`) indices
+The following definitions and `simp` lemmas are to allow any
+numeral-indexed element of a vector given with matrix notation to
+be extracted by `simp` (even when the numeral is larger than the
+number of elements in the vector, which is taken modulo that number
+of elements by virtue of the semantics of `bit0` and `bit1` and of
+addition on `fin n`).
+-/
+
+/-- `vec_join ho u v` joins two vectors of lengths `m` and `n` to produce
+one of length `o = m + n`.  `ho` provides control of definitional equality
+for the vector length. -/
+def vec_join (ho : o = m + n) (u : fin m → α) (v : fin n → α) : fin o → α :=
+λ i, if h : (i : ℕ) < m
+  then u ⟨i, h⟩
+  else v ⟨(i : ℕ) - m, (nat.sub_lt_left_iff_lt_add (le_of_not_lt h)).2 (ho ▸ i.property)⟩
+
+@[simp] lemma empty_join (v : fin n → α) : vec_join (zero_add _).symm ![] v = v :=
+by { ext, simp [vec_join] }
+
+@[simp] lemma cons_join (ho : o + 1 = m + 1 + n) (x : α) (u : fin m → α) (v : fin n → α) :
+  vec_join ho (vec_cons x u) v =
+    vec_cons x (vec_join (by rwa [add_assoc, add_comm 1, ←add_assoc,
+                                  add_right_cancel_iff] at ho) u v) :=
+begin
+  ext i,
+  simp_rw [vec_join],
+  split_ifs with h,
+  { rcases i with ⟨⟨⟩ | i, hi⟩,
+    { simp },
+    { simp only [nat.succ_eq_add_one, add_lt_add_iff_right, fin.coe_mk] at h,
+      simp [h] } },
+  { rcases i with ⟨⟨⟩ | i, hi⟩,
+    { simpa using h },
+    { rw [not_lt, fin.coe_mk, nat.succ_eq_add_one, add_le_add_iff_right] at h,
+      simp [h] } }
+end
+
+/-- `vec_alt0 v` gives a vector with half the length of `v`, with
+only alternate elements (even-numbered). -/
+def vec_alt0 (hm : m = n + n) (v : fin m → α) (k : fin n) : α :=
+v ⟨(k : ℕ) + k, hm.symm ▸ add_lt_add k.property k.property⟩
```

Comment:
Does it make sense to define this instead as:
```suggestion
def fin.bit0 (hm : m = n + n) (k : fin n) : fin m :=
\⟨(k : ℕ) + k, hm.symm ▸ add_lt_add k.property k.property⟩
```
and then using `v ∘ fin.bit0`?

Reply:
I'm not sure what other use there would be for that definition, but if we want it (possibly built on variants adding two different `fin` values, with or without `1` added as well) then presumably it goes in `data.fin`.

Reply:
Or perhaps for consistency with the other `bit0` functions,
```
def fin.bit0 {n : ℕ} (k : fin n) : fin (n + n) :=
⟨(k : ℕ) + k, add_lt_add k.property k.property⟩

/-- `hm` provides control of definitional equality -/
def fin.bit0' {m n : ℕ} (hm : m = n + n) : fin m := fin.cast hm.symm ∘ (fin.bit0 k)
```

Reply:
I'm mildly in favour of the `fin.bit0`/`fin.bit0'` proposal, but keeping `vec_alt0` until we really need `fin.bit0` also makes sense to me.

Reply:
If nothing else, `v ∘ fin.bit0` seems like a much clearer name to me than `vec_alt0`

Reply:
Also at that point, it seems quite desirable to express it as
```lean
def fin.bit0 {n : ℕ} (k : fin n) : fin (bit0 n) :=
⟨bit0 (k : ℕ), _⟩
```
to make the connection between `nat.bit0` and `fin.bit0` really clear.

Reply:
I don't see any `nat.bit0` definition, only `_root_.bit0`.

As noted above, using `bit0` and `bit1` in these definitions ends up meaning the subsequent proofs need to unfold `bit0` and `bit1` in several places; they're simpler if these definitions don't use `bit0` and `bit1`.

Reply:
How about as a compromise, we introduce `fin.bit0`, then as a follow-up I'll try re-expressing the proofs to use `bit0 k` instead of `k + k`?

Reply:
If you use `fin.bit0` for `vec_alt0`, then in the context of these lemmas you should have `fin.bit1` for `vec_alt1`; `vec_alt0` and `vec_alt1` are exactly analogous, extracting alternate elements of an even-length vector.

However, `fin.bit0` and `fin.bit1` would not be so clearly analogous to each other when seen on their own.  `fin.bit0` would map `fin n` to `fin (bit0 n)`, while `fin.bit1` would also map `fin n` to `fin (bit0 n)`; it would *not* map to `fin (bit1 n)`, which would be an odd inconsistency in the `fin` context. But if you make `fin.bit0` instead map to the smallest `fin` type that can hold all the required values, i.e. `fin (bit0 n - 1)`, that's no longer what's wanted for `vec_alt0`. And if you combine such a function with `cast_succ` to get back to `fin (bit0 n)` as needed for `vec_alt0`, you start needing to special-case `n = 0` because `fin 0` not `fin 1` is correct there.

I think of `bit0` and `bit1` as being about numerals; those definitions outside the context of numerals don't seem very helpful. I can see more possible use for two-argument versions that might be something like `fin.add_cast_add` and `find.add_add_one_cast_add` that map `fin m` and `fin n` to `fin (m + n)` (with an appropriate extra argument for control of definitional equality of types). That could e.g. map row and column indices in some context to indices for diagonals, or generally be relevant for some kind of convolution indexed by `fin` types. But it also wouldn't work with function composition for defining `vec_alt0` and `vec_alt1` in any such convenient way as `fin.bit0` or `fin.bit1`.